### PR TITLE
Bug fix/fix error handling

### PR DESCRIPTION
**Pull request by dothebart**
these 3 changes seem to be related to the python 3 migration, the parameter types aren't compatible. 

`sys.stderr.write(p_stderr)` will throw with: `TypeError: write() argument must be str, not bytes` ; the followup comment references python 2.5, raising the error gives the correct behaviour. 

I ran into this error by a gyp line spawning a process exiting errnously. 

The `None` as a parameter to eval `__builtins__` leads to the exception `Exception: 'NoneType' object is not subscriptable`.

### CODE REVIEW

**Comment by dothebart** [comment]
```  File "/home/willi/src/devel4/3rdParty/V8/v7.1.302.28/tools/gyp/pylib/gyp/generator/make.py", line 1779, in WriteMakeRule
    cmddigest = hashlib.sha1(command if command else self.target).hexdigest()
TypeError: Unicode-objects must be encoded before hashing
``` fixed by the 3rd commit.

File: gyp/pylib/gyp/generator/make.py [hunk 1]
```diff
@@ -1776,7 +1776,7 @@ def WriteMakeRule(self, outputs, inputs, actions=None, comment=None,
       # - The multi-output rule will have an do-nothing recipe.
 
       # Hash the target name to avoid generating overlong filenames.
-      cmddigest = hashlib.sha1(command if command else self.target).hexdigest()
+      cmddigest = hashlib.sha1(command.encode('utf-8') if command else self.target.encode('utf-8')).hexdigest()
```
Comment (cclauss) [inline gyp/pylib/gyp/generator/make.py]: `cmddigest = hashlib.sha1((command if command else self.target).encode('utf-8')).hexdigest()

Comment (dothebart) [reply gyp/pylib/gyp/generator/make.py]: check!

**Comment by cclauss** [comment]
I am +1 if we are sure these changes are compatible with both Py2 and Py3.

**Comment by dothebart** [comment]
doesn't error out with python 2 on debian testing.

**Comment by rvagg** [comment]
@dothebart what's the situation that's causing this error to be experienced? how could we go about reproducing it?

**Comment by dothebart** [comment]
@rvagg I was putting in the complete directory as `V8_ROOT` from the outside in, and it seems as if gyp can only handle relative paths for source files - at least with the makefile generator. I reverted to relative paths, and this is working now. 

But if its not supported to have absolute paths on filenames, gyp should rather error out; If I get it correctly it needs them relative to the location of the gyp-file - so 'stat' ing files, and ruling out absolute paths could keep people from running into misery.

**Comment by dothebart** [comment]
one of these issues was caused by a sub process erroring out - here due to non-python3-ness:
```
'<!<(PYTHON_EXECUTABLE) -c "import sys; print sys.byteorder ")',
```
so probably invoking `/bin/false` could reproduce it?

the other one was caused by this section in `features.gypi` for me: 
```
      [ 'component and "library" in component',
         {
        'is_component_build': 1,
      }, {
        'is_component_build': 0,
      }],
```

**Comment by rvagg** [comment]
sounds reasonable, @cclauss over to you I think

File: gyp/pylib/gyp/input.py [hunk 2]
```diff
@@ -911,9 +911,7 @@ def ExpandVariables(input, phase, variables, build_file):
           p_stdout, p_stderr = p.communicate('')
```
Comment (cclauss) [inline gyp/pylib/gyp/input.py]: For consistency, my preference would be the solution we use elsewhere:
```
if bytes != str:  # Python 3
    p_stdout = p_stdout.decode('utf-8')
    p_stderr = p_stderr.decode('utf-8')
```
And then revert the change to line 914.

Comment (cclauss) [reply gyp/pylib/gyp/input.py]: #1937

**Comment by rvagg** [comment]
@cclauss if you think this is important enough please take over the PR, submit a new one or whatever needs to be done so we can progress and get 6.0.1 out. If it's not that important then we'll just hold off on it.

**Comment by dothebart** [comment]
Sorry bit under stress @cclauss sent you an invite to my fork so you can make yourselves home ;)

**Comment by cclauss** [comment]
@rvagg Can you please rerun the one Travis CI job that timed out?

**Comment by rvagg** [comment]
landed in 1b11be6. I hope the commit msg is reasonable: `gyp: python3 fixes: utf8 decode, use of 'None' in eval`

**Comment by dothebart** [comment]
thank you!
